### DISCUSSION TITLE
How to sort or reorder columns selected with selectors

### DISCUSSION THREAD

**Discussion by ozgurkalan** [ibis-project/ibis#6795] [Q&A] [answered]
Hi there,

How to sort columns in this case:
`(t
.select(
s.startswith( ("H", "J") )
)
)`

**Comment by lostmygithubaccount** [comment, marked as answer]
you could re-select and sort the columns:

```python
t = t.select(sorted(t.columns))
```

and replace the sorted function with custom logic as needed

**Reply by lostmygithubaccount** [reply]
```python
[ins] In [1]: import ibis

[ins] In [2]: ibis.options.interactive = True

[ins] In [3]: t = ibis.examples.penguins.fetch()

[ins] In [4]: t = t.select(sorted(t.columns))

[ins] In [5]: t
Out[5]:
┏━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━━┳━━━━━━━┓
┃ bill_depth_mm ┃ bill_length_mm ┃ body_mass_g ┃ flipper_length_mm ┃ island    ┃ sex    ┃ species ┃ year  ┃
┡━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━╇━━━━━━━━━╇━━━━━━━┩
│ float64       │ float64        │ int64       │ int64             │ string    │ string │ string  │ int64 │
├───────────────┼────────────────┼─────────────┼───────────────────┼───────────┼────────┼─────────┼───────┤
│          18.7 │           39.1 │        3750 │               181 │ Torgersen │ male   │ Adelie  │  2007 │
│          17.4 │           39.5 │        3800 │               186 │ Torgersen │ female │ Adelie  │  2007 │
│          18.0 │           40.3 │        3250 │               195 │ Torgersen │ female │ Adelie  │  2007 │
│           nan │            nan │        NULL │              NULL │ Torgersen │ NULL   │ Adelie  │  2007 │
│          19.3 │           36.7 │        3450 │               193 │ Torgersen │ female │ Adelie  │  2007 │
│          20.6 │           39.3 │        3650 │               190 │ Torgersen │ male   │ Adelie  │  2007 │
│          17.8 │           38.9 │        3625 │               181 │ Torgersen │ female │ Adelie  │  2007 │
│          19.6 │           39.2 │        4675 │               195 │ Torgersen │ male   │ Adelie  │  2007 │
│          18.1 │           34.1 │        3475 │               193 │ Torgersen │ NULL   │ Adelie  │  2007 │
│          20.2 │           42.0 │        4250 │               190 │ Torgersen │ NULL   │ Adelie  │  2007 │
│             … │              … │           … │                 … │ …         │ …      │ …       │     … │
└───────────────┴────────────────┴─────────────┴───────────────────┴───────────┴────────┴─────────┴───────┘
```

**Reply by cpcloud** [reply]
We could consider adding a `Table.relocate` method, similar to [R's method of the same name](https://dplyr.tidyverse.org/reference/relocate.html).

**Reply by ozgurkalan** [reply]
working fine at the beginning as you have written, but very slow after filtering and using underscore.... Thanks a lot.

**Reply by cpcloud** [reply]
@ozgurkalan Can you share your code so that we might help you figure out what's slow?

**Reply by ozgurkalan** [reply]
No worries, previous works. It was exactly as below:
appr. 500k records with 59 cols
`
(t
.select(sorted(t.columns)) # this is added with your recommendation and works fast, in miliseconds
.select(sel.startswith(("H","J")))
.select(sorted(_.columns)) # this is still working at 5th minute and not responding yet
.order_by("H1")
.distinct()
)
`

**Reply by ozgurkalan** [reply]
by the way, a shape-like function would be excellent, I need to `t.to_pandas().shape` to get the shape, if there is any better way, really appreciate

**Reply by cpcloud** [reply]
I can make an issue for a `shape` property.

The cheapest way to compute `shape` of an arbitrary table expression is:

```python
def shape(expr) -> tuple[int, int]:
	return int(expr.count().execute()), len(expr.columns)
```

Note that `expr.count()` may not always be fast, but it will be much faster and more memory efficient than `expr.to_pandas().shape`.

**Reply by cpcloud** [reply]
> No worries, previous works. It was exactly as below: appr. 500k records with 59 cols `(t .select(sorted(t.columns)) # this is added with your recommendation and works fast, in miliseconds .select(sel.startswith(("H","J"))) .select(sorted(_.columns)) # this is still working at 5th minute and not responding yet .order_by("H1") .distinct() )`

`_.columns` is not doing what you think it does: `sorted(_.columns)` is going to run forever, because it uses `__getitem__` to access elements of a list. I'll put up a PR to make this fail fast.

To access `.columns` you may need to create a variable to hold the expression until we have a way to reorder columns with a `relocate` selector.
